--- /tmp/elf_compare/diff/nc_L1_TO_L1_unpack.insns	2026-05-20 22:23:51.377070246 +0000
+++ /tmp/elf_compare/diff/wc_L1_TO_L1_unpack.insns	2026-05-20 22:23:51.382070236 +0000
@@ -5,7 +5,7 @@
 lui	a5,0xffb00
 lui	a4,0xffb00
 addi	a5,a5,32
-addi	a4,a4,64
+addi	a4,a4,100
 bgeu	a5,a4,6044 <_start+0x44>
 addi	a4,a4,-1
 sub	a4,a4,a5
@@ -74,7 +74,7 @@
 sw	zero,-2004(gp)
 sw	zero,-2012(gp)
 sw	zero,-2016(gp)
-jal	6b50 <memset>
+jal	6cc4 <memset>
 lw	a2,4(s3)
 li	a5,1
 sw	a5,0(a2)
@@ -364,22 +364,62 @@
 seqz	a0,a1
 ret
 ret
-addi	sp,sp,-48
-sw	ra,44(sp)
-sw	s0,40(sp)
-sw	s1,36(sp)
-sw	s2,32(sp)
-sw	s3,28(sp)
-sw	s4,24(sp)
-sw	s5,20(sp)
-sw	s6,16(sp)
+addi	sp,sp,-64
+sw	s5,36(sp)
+addi	s5,gp,-1984
+lw	a5,0(s5)
+sw	ra,60(sp)
+sw	s0,56(sp)
+sw	s1,52(sp)
+sw	s2,48(sp)
+sw	s3,44(sp)
+sw	s4,40(sp)
+sw	s6,32(sp)
+sw	s7,28(sp)
+beqz	a5,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lui	a4,0x7c867
+lw	a3,4(s5)
+addi	a4,a4,-839
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,1
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,8(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,2
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,12(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,3
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,16(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,4
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,20(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a4,5
+beq	a5,a4,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lui	a4,0x7c867
+lw	a3,24(s5)
+addi	a4,a4,-839
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,6
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,28(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a3,7
+beq	a5,a3,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
+lw	a3,32(s5)
+beq	a3,a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+li	a4,8
+bne	a5,a4,6b8c <run_kernel(RuntimeParams const&)+0x5d4>
 sb	zero,0(sp)
 lw	a5,-2012(gp)
 lw	a3,-2016(gp)
 li	a2,3
 add	a4,a5,a3
 addi	a4,a4,-1021
-bgeu	a2,a4,6640 <run_kernel(RuntimeParams const&)+0x88>
+bgeu	a2,a4,66e0 <run_kernel(RuntimeParams const&)+0x128>
 lui	a4,0xffb12
 lw	a1,496(a4)
 lw	a4,504(a4)
@@ -402,44 +442,44 @@
 mv	a0,a1
 li	a2,1
 jal	62dc <ckernel::unpacker::is_unpacker_format_conversion_supported_fp32_acc(DataFormat, DataFormat, bool)>
-beqz	a0,6afc <run_kernel(RuntimeParams const&)+0x544>
+beqz	a0,6c14 <run_kernel(RuntimeParams const&)+0x65c>
 lw	a4,-2036(gp)
 lw	a5,52(a4)
 zext.b	a5,a5
-bnez	a5,6658 <run_kernel(RuntimeParams const&)+0xa0>
+bnez	a5,66f8 <run_kernel(RuntimeParams const&)+0x140>
 ttsetadcxy	3,0,0,0,0,11
 ttsetadczw	3,0,0,0,0,15
 lw	a4,-2000(gp)
 lui	a5,0xffef0
-beqz	a4,667c <run_kernel(RuntimeParams const&)+0xc4>
+beqz	a4,671c <run_kernel(RuntimeParams const&)+0x164>
 addi	a5,a5,896
 li	a4,512
 sw	a4,228(a5)
 sw	a4,236(a5)
 ttatgetm	0
-lui	s5,0xffe40
-mv	s5,s5
+lui	s6,0xffe40
+mv	s6,s6
 lui	a4,0xb3ff0
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb47f0
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb3070
 addi	a4,a4,1
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb4800
 addi	a4,a4,1
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb5010
 addi	a4,a4,1
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb3010
 addi	a4,a4,2
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0xb5400
 addi	a3,a4,71
-sw	a3,0(s5)
+sw	a3,0(s6)
 addi	a4,a4,119
-sw	a4,0(s5)
+sw	a4,0(s6)
 ttatrelm	0
 li	a4,21
 sw	a4,256(a5)
@@ -459,11 +499,11 @@
 sw	a4,484(a5)
 lui	a4,0x5e240
 addi	a4,a4,-1024
-sw	a4,0(s5)
+sw	a4,0(s6)
 lui	a4,0x5e440
 addi	a4,a4,-1024
 lui	a2,0x400
-sw	a4,0(s5)
+sw	a4,0(s6)
 addi	a2,a2,64
 sw	a2,336(a5)
 lw	a4,-2040(gp)
@@ -490,34 +530,27 @@
 ttsetc16	41,0
 lui	a5,0x45000
 addi	a3,a5,72
-sw	a3,0(s5)
+sw	a3,0(s6)
 lui	a4,0x2021
 addi	a5,a5,74
 addi	a4,a4,16
-sw	a5,0(s5)
+sw	a5,0(s6)
 addi	a0,sp,4
 sw	a4,4(sp)
 jal	627c <ckernel::validate_tensor_shape_tile_dependent_ops_(ckernel::TensorShape const&)>
-beqz	a0,6acc <run_kernel(RuntimeParams const&)+0x514>
+beqz	a0,6c20 <run_kernel(RuntimeParams const&)+0x668>
 lui	a5,0xb4010
 addi	a5,a5,72
-sw	a5,0(s5)
-li	a5,16
-li	a4,4
-bltu	a4,a5,6a6c <run_kernel(RuntimeParams const&)+0x4b4>
-li	a4,1
-beq	a5,a4,6a7c <run_kernel(RuntimeParams const&)+0x4c4>
-li	a4,2
-bne	a5,a4,6aec <run_kernel(RuntimeParams const&)+0x534>
-ttsetadcxx	3,31,0
+sw	a5,0(s6)
+ttsetadcxx	3,255,0
 lw	a5,4(sp)
 addi	a0,sp,12
-srli	s6,a5,0x10
+srli	s7,a5,0x10
 sw	a5,12(sp)
-zext.b	s6,s6
-srli	s5,a5,0x18
+zext.b	s7,s7
+srli	s6,a5,0x18
 jal	627c <ckernel::validate_tensor_shape_tile_dependent_ops_(ckernel::TensorShape const&)>
-beqz	a0,6af4 <run_kernel(RuntimeParams const&)+0x53c>
+beqz	a0,6c0c <run_kernel(RuntimeParams const&)+0x654>
 lw	a5,-2036(gp)
 li	a3,0
 addi	a5,a5,8
@@ -525,8 +558,8 @@
 lw	a3,0(a5)
 and	zero,zero,a3
 lui	a5,0xffb80
-sw	s6,0(a5)
-sw	s5,4(a5)
+sw	s7,0(a5)
+sw	s6,4(a5)
 lui	a3,0x2000
 sw	a3,8(a5)
 sw	a3,12(a5)
@@ -546,7 +579,7 @@
 lw	a5,0(a4)
 and	zero,zero,a5
 lbu	a5,0(sp)
-beqz	a5,68dc <run_kernel(RuntimeParams const&)+0x324>
+beqz	a5,6960 <run_kernel(RuntimeParams const&)+0x3a8>
 lui	a5,0xffb12
 lw	a1,496(a5)
 lw	a5,504(a5)
@@ -565,13 +598,51 @@
 sw	a1,4(a3)
 sw	a4,-2012(gp)
 sw	a5,-2016(gp)
+lw	a5,0(s5)
+beqz	a5,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lui	a4,0xbd77
+lw	a3,4(s5)
+addi	a4,a4,-244
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,1
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,8(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,2
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,12(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,3
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,16(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,4
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,20(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a4,5
+beq	a5,a4,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lui	a4,0xbd77
+lw	a3,24(s5)
+addi	a4,a4,-244
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,6
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,28(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a3,7
+beq	a5,a3,6bf0 <run_kernel(RuntimeParams const&)+0x638>
+lw	a3,32(s5)
+beq	a3,a4,69f8 <run_kernel(RuntimeParams const&)+0x440>
+li	a4,8
+bne	a5,a4,6bf0 <run_kernel(RuntimeParams const&)+0x638>
 sb	zero,12(sp)
 lw	a5,-2012(gp)
 lw	a3,-2016(gp)
 li	a2,3
 add	a4,a3,a5
 addi	a4,a4,-1021
-bgeu	a2,a4,6940 <run_kernel(RuntimeParams const&)+0x388>
+bgeu	a2,a4,6a5c <run_kernel(RuntimeParams const&)+0x4a4>
 lui	a4,0xffb12
 lw	a1,496(a4)
 lw	a4,504(a4)
@@ -608,35 +679,35 @@
 ttsetadczw	3,0,0,0,0,15
 lw	a5,-2000(gp)
 lui	a0,0xffef0
-beqz	a5,6990 <run_kernel(RuntimeParams const&)+0x3d8>
+beqz	a5,6aac <run_kernel(RuntimeParams const&)+0x4f4>
 addi	a0,t0,896
 lw	a5,52(a4)
 andi	a5,a5,254
-bnez	a5,6990 <run_kernel(RuntimeParams const&)+0x3d8>
-bnez	a1,6a84 <run_kernel(RuntimeParams const&)+0x4cc>
+bnez	a5,6aac <run_kernel(RuntimeParams const&)+0x4f4>
+bnez	a1,6ba8 <run_kernel(RuntimeParams const&)+0x5f0>
 sw	a7,304(a0)
 sw	a3,496(a0)
 lw	a5,52(a4)
 mv	a3,a1
 zext.b	a5,a5
-bltu	t6,a5,6a9c <run_kernel(RuntimeParams const&)+0x4e4>
+bltu	t6,a5,6bc0 <run_kernel(RuntimeParams const&)+0x608>
 sw	zero,52(a4)
 ttstallwait	8,1024
 ttmop	1,0,0
 ttsemget	32
 sub	a1,a6,a3
 sw	a1,-1988(gp)
-beq	a3,a6,6ac4 <run_kernel(RuntimeParams const&)+0x50c>
+beq	a3,a6,6be8 <run_kernel(RuntimeParams const&)+0x630>
 ttsetc16	41,257
 add	a2,a2,t4
-bne	a2,t5,6970 <run_kernel(RuntimeParams const&)+0x3b8>
+bne	a2,t5,6a8c <run_kernel(RuntimeParams const&)+0x4d4>
 li	a5,0
 addi	a4,a4,4
 sw	a5,0(a4)
 lw	a5,0(a4)
 and	zero,zero,a5
 lbu	a5,12(sp)
-beqz	a5,6a44 <run_kernel(RuntimeParams const&)+0x48c>
+beqz	a5,6b60 <run_kernel(RuntimeParams const&)+0x5a8>
 lui	a5,0xffb12
 lw	a1,496(a5)
 lw	a5,504(a5)
@@ -655,28 +726,30 @@
 sw	a1,4(a3)
 sw	a4,-2012(gp)
 sw	a5,-2016(gp)
-lw	ra,44(sp)
-lw	s0,40(sp)
-lw	s1,36(sp)
-lw	s2,32(sp)
-lw	s3,28(sp)
-lw	s4,24(sp)
-lw	s5,20(sp)
-lw	s6,16(sp)
-addi	sp,sp,48
-ret
-li	a4,8
-bne	a5,a4,6aec <run_kernel(RuntimeParams const&)+0x534>
-ttsetadcxx	3,127,0
-j	6800 <run_kernel(RuntimeParams const&)+0x248>
-ttsetadcxx	3,15,0
-j	6800 <run_kernel(RuntimeParams const&)+0x248>
+lw	ra,60(sp)
+lw	s0,56(sp)
+lw	s1,52(sp)
+lw	s2,48(sp)
+lw	s3,44(sp)
+lw	s4,40(sp)
+lw	s5,36(sp)
+lw	s6,32(sp)
+lw	s7,28(sp)
+addi	sp,sp,64
+ret
+lui	a4,0x7c867
+sh2add	a3,a5,s5
+addi	a4,a4,-839
+addi	a5,a5,1
+sw	a4,4(a3)
+sw	a5,0(s5)
+j	667c <run_kernel(RuntimeParams const&)+0xc4>
 sw	a7,308(a0)
 sw	a3,500(a0)
 lw	a5,52(a4)
 mv	a3,a1
 zext.b	a5,a5
-bgeu	t6,a5,69b8 <run_kernel(RuntimeParams const&)+0x400>
+bgeu	t6,a5,6ad4 <run_kernel(RuntimeParams const&)+0x51c>
 ebreak
 lw	a4,-2036(gp)
 lw	a3,-1988(gp)
@@ -686,34 +759,54 @@
 ttsemget	32
 sub	a1,a6,a3
 sw	a1,-1988(gp)
-bne	a3,a6,69d4 <run_kernel(RuntimeParams const&)+0x41c>
+bne	a3,a6,6af0 <run_kernel(RuntimeParams const&)+0x538>
 ttsetc16	41,0
-j	69d8 <run_kernel(RuntimeParams const&)+0x420>
+j	6af4 <run_kernel(RuntimeParams const&)+0x53c>
+lui	a4,0xbd77
+sh2add	a3,a5,s5
+addi	a4,a4,-244
+addi	a5,a5,1
+sw	a4,4(a3)
+sw	a5,0(s5)
+j	69f8 <run_kernel(RuntimeParams const&)+0x440>
+ebreak
+j	68a4 <run_kernel(RuntimeParams const&)+0x2ec>
+ebreak
+ebreak
+j	66f4 <run_kernel(RuntimeParams const&)+0x13c>
 ebreak
 lui	a4,0xb4010
 addi	a4,a4,72
 lbu	a5,4(sp)
-sw	a4,0(s5)
+sw	a4,0(s6)
 li	a4,16
-bgeu	a4,a5,6b08 <run_kernel(RuntimeParams const&)+0x550>
+bgeu	a4,a5,6c44 <run_kernel(RuntimeParams const&)+0x68c>
 ebreak
-ttsetadcxx	3,255,0
-j	6800 <run_kernel(RuntimeParams const&)+0x248>
-ebreak
-j	6820 <run_kernel(RuntimeParams const&)+0x268>
-ebreak
-ebreak
-j	6654 <run_kernel(RuntimeParams const&)+0x9c>
+j	6880 <run_kernel(RuntimeParams const&)+0x2c8>
 lui	a4,0xffff0
 addi	a4,a4,-279
 sra	a4,a4,a5
 andi	a4,a4,1
-beqz	a4,6b20 <run_kernel(RuntimeParams const&)+0x568>
-ebreak
+bnez	a4,6c8c <run_kernel(RuntimeParams const&)+0x6d4>
 li	a4,4
-bne	a5,a4,67e4 <run_kernel(RuntimeParams const&)+0x22c>
+beq	a5,a4,6c7c <run_kernel(RuntimeParams const&)+0x6c4>
+bltu	a4,a5,6c94 <run_kernel(RuntimeParams const&)+0x6dc>
+li	a4,1
+beq	a5,a4,6c84 <run_kernel(RuntimeParams const&)+0x6cc>
+li	a4,2
+bne	a5,a4,6880 <run_kernel(RuntimeParams const&)+0x2c8>
+ttsetadcxx	3,31,0
+j	6884 <run_kernel(RuntimeParams const&)+0x2cc>
 ttsetadcxx	3,63,0
-j	6800 <run_kernel(RuntimeParams const&)+0x248>
+j	6884 <run_kernel(RuntimeParams const&)+0x2cc>
+ttsetadcxx	3,15,0
+j	6884 <run_kernel(RuntimeParams const&)+0x2cc>
+ebreak
+j	6c58 <run_kernel(RuntimeParams const&)+0x6a0>
+li	a4,8
+bne	a5,a4,6880 <run_kernel(RuntimeParams const&)+0x2c8>
+ttsetadcxx	3,127,0
+j	6884 <run_kernel(RuntimeParams const&)+0x2cc>
 ret
 ret
 lui	a5,0x20
@@ -724,10 +817,10 @@
 ret
 li	t1,15
 mv	a4,a0
-bgeu	t1,a2,6b94 <memset+0x44>
+bgeu	t1,a2,6d08 <memset+0x44>
 andi	a5,a4,15
-bnez	a5,6c00 <memset+0xb0>
-bnez	a1,6be8 <memset+0x98>
+bnez	a5,6d74 <memset+0xb0>
+bnez	a1,6d5c <memset+0x98>
 andi	a3,a2,-16
 andi	a2,a2,15
 add	a3,a3,a4
@@ -736,8 +829,8 @@
 sw	a1,8(a4)
 sw	a1,12(a4)
 addi	a4,a4,16
-bltu	a4,a3,6b74 <memset+0x24>
-bnez	a2,6b94 <memset+0x44>
+bltu	a4,a3,6ce8 <memset+0x24>
+bnez	a2,6d08 <memset+0x44>
 ret
 sub	a3,t1,a2
 slli	a3,a3,0x2
@@ -765,7 +858,7 @@
 or	a1,a1,a3
 slli	a3,a1,0x10
 or	a1,a1,a3
-j	6b68 <memset+0x18>
+j	6cdc <memset+0x18>
 slli	a3,a5,0x2
 auipc	t0,0x0
 add	a3,a3,t0
@@ -775,5 +868,5 @@
 addi	a5,a5,-16
 sub	a4,a4,a5
 add	a2,a2,a5
-bgeu	t1,a2,6b94 <memset+0x44>
-j	6b64 <memset+0x14>
+bgeu	t1,a2,6d08 <memset+0x44>
+j	6cd8 <memset+0x14>
