--- a/include/ATen/Dispatch.h	2025-09-22 15:28:21.557819366 +0800
+++ b/include/ATen/Dispatch.h	2025-09-22 15:36:25.390220440 +0800
@@ -95,13 +95,13 @@
     return __VA_ARGS__();                                                   \
   }

-namespace detail {
+namespace disp_detail {

 inline at::ScalarType scalar_type(at::ScalarType s) {
   return s;
 }

-} // namespace detail
+} // namespace disp_detail

 // The AT_DISPATCH_* family of macros provides the ability to
 // conveniently generate specializations of a kernel over all of the
@@ -195,7 +195,7 @@
     const auto& the_type = TYPE;                                            \
     constexpr const char* at_dispatch_name = NAME;                          \
     /* don't use TYPE again in case it is an expensive or side-effect op */ \
-    at::ScalarType _st = ::detail::scalar_type(the_type);                   \
+    at::ScalarType _st = ::disp_detail::scalar_type(the_type);                   \
     RECORD_KERNEL_FUNCTION_DTYPE(at_dispatch_name, _st);                    \
     switch (_st) {                                                          \
       __VA_ARGS__                                                           \
