diff --git a/src/nb_ndarray.cpp b/src/nb_ndarray.cpp
index 1c25e41..d005cf2 100644
--- a/src/nb_ndarray.cpp
+++ b/src/nb_ndarray.cpp
@@ -524,6 +524,7 @@ ndarray_handle *ndarray_import(PyObject *o, const ndarray_config *c,
                 case (uint8_t) dlpack::dtype_code::Int: prefix = "int"; break;
                 case (uint8_t) dlpack::dtype_code::UInt: prefix = "uint"; break;
                 case (uint8_t) dlpack::dtype_code::Float: prefix = "float"; break;
+                case (uint8_t) dlpack::dtype_code::Bfloat: prefix = "bfloat"; break;
                 case (uint8_t) dlpack::dtype_code::Complex: prefix = "complex"; break;
                 default:
                     return nullptr;
