diff --git a/CMakeLists.txt b/CMakeLists.txt
index e73c8fd..09248ef 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -86,7 +86,6 @@ get_filename_component(TOKENIZERS_CPP_ROOT ${CMAKE_CURRENT_LIST_FILE} DIRECTORY)
 set(TOKENIZERS_CPP_CARGO_SOURCE_PATH ${TOKENIZERS_CPP_ROOT}/rust)

 option(MSGPACK_USE_BOOST "Use Boost libraried" OFF)
-add_subdirectory(msgpack)

 option(MLC_ENABLE_SENTENCEPIECE_TOKENIZER "Enable SentencePiece tokenizer" OFF)

diff --git a/sentencepiece/CMakeLists.txt b/sentencepiece/CMakeLists.txt
index a2f0f77..9fb7b4f 100644
--- a/sentencepiece/CMakeLists.txt
+++ b/sentencepiece/CMakeLists.txt
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.!

-cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.1...3.30 FATAL_ERROR)
 file(STRINGS "VERSION.txt" SPM_VERSION)
 message(STATUS "VERSION: ${SPM_VERSION}")
