diff --git a/CMakeLists.txt b/CMakeLists.txt
index 796dc46..d6ce609 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,11 +5,11 @@
 # Distributed under the terms of the BSD 3-Clause License.                 #
 #                                                                          #
 # The full license is in the file LICENSE, distributed with this software. #
 ############################################################################

-cmake_minimum_required(VERSION 3.29)
+cmake_minimum_required(VERSION 3.15..3.30)
 project(xtl)

 set(XTL_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/include)

 # Versioning
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index ee6ce9c..ac56750 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -5,11 +5,11 @@
 # Distributed under the terms of the BSD 3-Clause License.                 #
 #                                                                          #
 # The full license is in the file LICENSE, distributed with this software. #
 ############################################################################

-cmake_minimum_required(VERSION 3.29)
+cmake_minimum_required(VERSION 3.15..3.30)

 find_package(doctest            REQUIRED)

 if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
     project(xtl-test)
--
2.49.0
