diff --git a/c++/src/kj/async-inl.h b/c++/src/kj/async-inl.h
index fc8a9813..c5c78922 100644
--- a/c++/src/kj/async-inl.h
+++ b/c++/src/kj/async-inl.h
@@ -203,6 +203,8 @@ public:
   String traceEvent();
   // Helper that builds a trace and stringifies it.
 
+  void disableFiring() { firing = false; }
+
 protected:
   virtual Maybe<Own<Event>> fire() = 0;
   // Fire the event.  Possibly returns a pointer to itself, which will be discarded by the
diff --git a/c++/src/kj/async.c++ b/c++/src/kj/async.c++
index 7c76f1d5..428d4eee 100644
--- a/c++/src/kj/async.c++
+++ b/c++/src/kj/async.c++
@@ -342,6 +342,7 @@ protected:
 
     // Call the error handler if there was an exception.
     KJ_IF_MAYBE(e, result.exception) {
+      self->disableFiring();
       taskSet.errorHandler.taskFailed(kj::mv(*e));
     }
 
