diff --git a/src/nb_ndarray.cpp b/src/nb_ndarray.cpp
index 03c4492..3d38b73 100644
--- a/src/nb_ndarray.cpp
+++ b/src/nb_ndarray.cpp
@@ -702,10 +702,13 @@ ndarray_handle *ndarray_import(PyObject *src, const ndarray_config *c,
                     prefix = "uint";
                     break;
                 case (uint8_t) dlpack::dtype_code::Float:
                     prefix = "float";
                     break;
+                case (uint8_t) dlpack::dtype_code::Bfloat:
+                    prefix = "bfloat";
+                    break;
                 case (uint8_t) dlpack::dtype_code::Complex:
                     prefix = "complex";
                     break;
                 default:
                     return nullptr;
--
2.52.0
